x-pack/plugins/enterprise_search/server/routes/enterprise_search/indices.ts

```diff
@@ -989,4 +994,165 @@ export function registerIndexRoutes({
       }
     })
   );
+
+  router.post(
+    {
+      path: '/internal/enterprise_search/ml/models/{modelName}',
+      validate: {
+        params: schema.object({
+          modelName: schema.string(),
+        }),
+      },
+    },
+    elasticsearchErrorHandler(log, async (context, request, response) => {
+      const modelName = decodeURIComponent(request.params.modelName);
+      const {
+        elasticsearch: { client },
+        savedObjects: { client: savedObjectsClient },
+      } = await context.core;
+      const trainedModelsProvider = ml
+        ? await ml.trainedModelsProvider(request, savedObjectsClient)
+        : undefined;
+
+      const savedObjectService = mlSavedObjectServiceFactory(
+        savedObjectsClient,
+        savedObjectsClient,
+        true,
+        undefined,
+        client,
+        () => Promise.resolve() // pretend isMlReady, to allow us to initialize the saved objects
+      );
+
+      const mlClient = savedObjectService
+        ? await getMlClient(client, savedObjectService)
+        : undefined;
+
+      try {
+        const deployResult = await startMlModelDownload(
+          modelName,
+          client,
+          mlClient,
+          trainedModelsProvider,
+          savedObjectService
+        );
+
+        return response.ok({
+          body: deployResult,
+          headers: { 'content-type': 'application/json' },
+        });
+      } catch (error) {
+        if (isResourceNotFoundException(error)) {
+          // return specific message if model doesn't exist
+          return createError({
+            errorCode: ErrorCode.RESOURCE_NOT_FOUND,
+            message: error.meta?.body?.error?.reason,
+            response,
+            statusCode: 404,
+          });
+        }
+        // otherwise, let the default handler wrap it
+        throw error;
+      }
+    })
+  );
+
+  router.post(
+    {
+      path: '/internal/enterprise_search/ml/models/{modelName}/deploy',
+      validate: {
+        params: schema.object({
+          modelName: schema.string(),
+        }),
+      },
+    },
+    elasticsearchErrorHandler(log, async (context, request, response) => {
+      const modelName = decodeURIComponent(request.params.modelName);
+      const {
+        elasticsearch: { client },
+        savedObjects: { client: savedObjectsClient },
+      } = await context.core;
+      const trainedModelsProvider = ml
+        ? await ml.trainedModelsProvider(request, savedObjectsClient)
+        : undefined;
+
+      const savedObjectService = mlSavedObjectServiceFactory(
```

Comment:
We're doing the same (or very similar) initialization and wrapping to check a 404 error in all three endpoints. Can we generalize this to a separate wrapping function?

Reply:
good call on these... I didn't wrap things, but I did clean up and have a common function to check this